roles/tests-integration/tests/common/mod.rs

```diff
@@ -147,8 +158,129 @@ impl TemplateProvider {
+            cb_outs
+        } else {
+            vec![CoinbaseOutput::new(
+                "P2WPKH".to_string(),
+                "036adc3bdf21e6f9a0f0fb0066bf517e5b7909ed1563d6958a10993849a7554075".to_string(),
+            )]
+        };
+        let pool_signature = "Stratum v2 SRI Pool".to_string();
+        let tp_address = if let Some(tp_add) = template_provider_address {
+            tp_add.to_string()
+        } else {
+            "127.0.0.1:8442".to_string()
+        };
+        let connection_config = pool_sv2::mining_pool::ConnectionConfig::new(
+            listening_address.to_string(),
+            cert_validity_sec,
+            pool_signature,
+        );
+        let template_provider_config =
+            pool_sv2::mining_pool::TemplateProviderConfig::new(tp_address, None);
+        let authority_config =
+            pool_sv2::mining_pool::AuthorityConfig::new(authority_public_key, authority_secret_key);
+        let config = Configuration::new(
+            connection_config,
+            template_provider_config,
+            authority_config,
+            coinbase_outputs,
+        );
+        let pool = PoolSv2::new(config);
+
+        Self {
+            pool,
+            port: pool_port,
+        }
+    }
+}
+
+pub async fn start_template_provider() -> (TemplateProvider, u16) {
+    let template_provider_port = get_available_port();
+    let template_provider = TemplateProvider::start(template_provider_port);
+    template_provider.generate_blocks(16);
+    (template_provider, template_provider_port)
+}
 
-    pub fn get_block_count(&self) -> u64 {
-        self.bitcoind.client.get_block_count().unwrap()
+pub async fn start_template_provider_and_pool() -> Result<(PoolSv2, u16, TemplateProvider, u16), ()>
+{
+    let (template_provider, template_provider_port) = start_template_provider().await;
+    let template_provider_address =
+        SocketAddr::from_str(&format!("127.0.0.1:{}", template_provider_port)).unwrap();
+    let test_pool = TestPoolSv2::new(None, None, Some(template_provider_address));
+    let pool = test_pool.pool.clone();
+    let state = pool.state().await.safe_lock(|s| s.clone()).unwrap();
+    assert_eq!(state, pool_sv2::PoolState::Initial);
+    let _pool = pool.clone();
+    tokio::task::spawn(async move {
+        assert!(_pool.start().await.is_ok());
+    });
+    // Wait for the pool to start.
+    tokio::time::sleep(Duration::from_secs(1)).await;
+    let pool_listening_address =
+        SocketAddr::from_str(&format!("127.0.0.1:{}", test_pool.port)).unwrap();
+    loop {
+        if is_port_open(pool_listening_address) {
+            break;
+        }
     }
+    let state = pool.state().await.safe_lock(|s| s.clone()).unwrap();
+    assert_eq!(state, pool_sv2::PoolState::Running);
+    template_provider.stop();
+    Ok((
+        pool,
+        test_pool.port,
+        template_provider,
+        template_provider_port,
+    ))
```

Comment:
Why are we having a method to spun up both template provider and pool together?

Reply:
The main motivation here is to provide users(devs) with a single function to start both the TP and a Pool so they don't need to rewrite all of this code again and again

Reply:
while this seems convenient at a first glance, I wonder how it will scale?

if we think that there's 7 different roles, and there could be scenarios where we want more than one instance (e.g.: 2 proxies connected to the same pool), the combinatorial possibilities can explode very quickly

Reply:
This should more useful in testing scenarios that involve the `pool`, i.e., the MG tests that are prefixed with `pool_`. All the `pool` tests need to start a TP before starting the `pool` so I wanted to avoid duplicating the code. 

An example can be seen here: https://github.com/stratum-mining/stratum/pull/1164/files#diff-2a0b80af72100f3ef6475d64910e57bfa2194a668b3a7692391a53f964340fb4R68
all the code from L38 to L68 can be removed and the test will be much nicer.

Reply:
Also, once we use it in a test(in #1164 for example), we can remove [this](https://github.com/stratum-mining/stratum/pull/1066/files#diff-2a0b80af72100f3ef6475d64910e57bfa2194a668b3a7692391a53f964340fb4R6) test that is ONLY calling this function

Reply:
Following this logic, we would need to implement similar methods for JDC and TP setup as well. This approach becomes problematic when setting up a full Job Declarator configuration, as both methods would be ineffective due to the need for TP instances to use different ports. For each test, we may end up with some duplicated setup code, which could serve as a framework.

Reply:
Maybe something like this?

```suggestion
pub async fn start_template_provider_and_pool(tp_port: Option<..>) -> Result<(PoolSv2, u16, TemplateProvider, u16), ()>
{
    let (template_provider, template_provider_port) = start_template_provider(tp_port).await;
    let template_provider_address =
        SocketAddr::from_str(&format!("127.0.0.1:{}", template_provider_port)).unwrap();
    let test_pool = TestPoolSv2::new(None, None, Some(template_provider_address));
    let pool = test_pool.pool.clone();
    let state = pool.state().await.safe_lock(|s| s.clone()).unwrap();
    assert_eq!(state, pool_sv2::PoolState::Initial);
    let _pool = pool.clone();
    tokio::task::spawn(async move {
        assert!(_pool.start().await.is_ok());
    });
    // Wait for the pool to start.
    tokio::time::sleep(Duration::from_secs(1)).await;
    let pool_listening_address =
        SocketAddr::from_str(&format!("127.0.0.1:{}", test_pool.port)).unwrap();
    loop {
        if is_port_open(pool_listening_address) {
            break;
        }
    }
    let state = pool.state().await.safe_lock(|s| s.clone()).unwrap();
    assert_eq!(state, pool_sv2::PoolState::Running);
    template_provider.stop();
    Ok((
        pool,
        test_pool.port,
        template_provider,
        template_provider_port,
    ))
```

And you are right, at the beginning there is a bit of overhead needing to add the different `start` functions

Reply:
I am still not convinced about this. The example I provided was just one specific case. If, for instance, we want to change other configuration properties of any of the involved entities, it would require further modifications to the function signature. All functions should adhere to the Open/Closed Principle: they should be open for extension but closed to modification.